corehq/blobs/mixin.py

```diff
@@ -96,30 +138,44 @@ def put_attachment(self, content, name=None, content_type=None, content_length=N
             name = getattr(content, "name", None)
         if name is None:
             raise InvalidAttachment("cannot save attachment without name")
+        if self._id is None:
+            raise ResourceNotFound("cannot put attachment on unidentified document")
+        if hasattr(self, "domain"):
```

Comment:
Why is this check needed?

Reply:
Some objects that use `put_attachment` do not have a `domain` attribute. Example: `InvoicePdf`

https://github.com/dimagi/commcare-hq/blob/25d83a6250abcbde256ef275ef2145f237929873/corehq/apps/accounting/models.py#L2883

The blob db now requires a domain when saving blobs. There are two ways the domain may be provided here:

- `domain` argument passed to this method (must match `self.domain` if the attribute exists).
- no `domain` passed -> use `self.domain` value (if present), otherwise raise an error.